File: hto/05/hto_05.py
from typing import List

class Solution:
    def replaceSpace(self, s: str) -> str:
        res = []
        for c in s:
            if c == ' ':
                res.append("%20")
            else:
                res.append(c)
        return ''.join(res)

    def replaceSpace2(self, s: str) -> str:
        return ''.join(("%20" if c == ' ' else c for c in s))


if __name__ == "__main__":
    s = Solution()
    print(s.replaceSpace("We are happy"))
    print(s.replaceSpace2("You are sad"))

File: hto/05/hto_05.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

char *replaceSpace(char *s){
    size_t i = strlen(s);
    size_t b = 0;
    for (int k = 0; s[k]; ++k) {
        if (s[k] == ' ') b++;
    }
    /*
    while (*s) {
        if (*s == ' ') b++;
        s++;   // s will become empty string
    }
     */

    size_t j = i + b * 2;
    char *res = (char *)malloc(sizeof(char) * (j + 1));
    res[j] = '\0';
    while (i) {
        if (s[i - 1] == ' ') {
            res[--j] = '0';
            res[--j] = '2';
            res[--j] = '%';
        } else {
            res[--j] = s[i - 1];
        }
        i--;
    }
    return res;
}

int main(int argc, char *argv[])
{
    char *s = "We are happy.";
    printf("%s\n", replaceSpace(s));
}